### DISCUSSION TITLE
Error in  Exact Diagonalization of User Modified Quantum Spin hamiltonian

### DISCUSSION THREAD

**Discussion by magnon00** [QuSpin/QuSpin#454] [Q&A] [answered]
I have modified the spin hamiltonian  xxz chain model 

<a href="https://www.codecogs.com/eqnedit.php?latex=H&space;=&space;\sum_{j=0}^{L-2}\frac{J_{xy}}{2}\left(S^&plus;_{j&plus;1}S^-_{j}&space;&plus;&space;\mathrm{h.c.}\right)&space;&plus;&space;J_{zz}S^z_{j&plus;1}S^z_{j}&space;&plus;&space;h_z\sum_{j=0}^{L-1}S^z_{j}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?H&space;=&space;\sum_{j=0}^{L-2}\frac{J_{xy}}{2}\left(S^&plus;_{j&plus;1}S^-_{j}&space;&plus;&space;\mathrm{h.c.}\right)&space;&plus;&space;J_{zz}S^z_{j&plus;1}S^z_{j}&space;&plus;&space;h_z\sum_{j=0}^{L-1}S^z_{j}" title="H = \sum_{j=0}^{L-2}\frac{J_{xy}}{2}\left(S^+_{j+1}S^-_{j} + \mathrm{h.c.}\right) + J_{zz}S^z_{j+1}S^z_{j} + h_z\sum_{j=0}^{L-1}S^z_{j}" /></a>

given in following tutorial: 

[QuSpin's exact  diagonlization routines to solve for the eigenstates and energies of the XXZ chain.](https://github.com/weinbe58/QuSpin/blob/master/examples/scripts/example0.py)

to 

xyz chain model.     

<a href="https://www.codecogs.com/eqnedit.php?latex=H&space;=&space;\sum_{j=0}^{L-2}J_{xx}S^x_{j&plus;1}S^x_{j}&space;&plus;&space;J_{yy}S^y_{j&plus;1}S^y_{j}&space;&plus;&space;J_{zz}S^z_{j&plus;1}S^z_{j}&space;&plus;&space;h_z\sum_{j=0}^{L-1}S^z_{j}," target="_blank"><img src="https://latex.codecogs.com/png.latex?H&space;=&space;\sum_{j=0}^{L-2}J_{xx}S^x_{j&plus;1}S^x_{j}&space;&plus;&space;J_{yy}S^y_{j&plus;1}S^y_{j}&space;&plus;&space;J_{zz}S^z_{j&plus;1}S^z_{j}&space;&plus;&space;h_z\sum_{j=0}^{L-1}S^z_{j}," title="H = \sum_{j=0}^{L-2}J_{xx}S^x_{j+1}S^x_{j} + J_{yy}S^y_{j+1}S^y_{j} + J_{zz}S^z_{j+1}S^z_{j} + h_z\sum_{j=0}^{L-1}S^z_{j}," /></a>

`from __future__ import print_function, division`
`import sys,os`
`from quspin.operators import hamiltonian` 
`from quspin.basis import spin_basis_1d` 
`import numpy as np `
`hz=2.0`
`L=10 `
`Jxx=2.0` 
`Jyy=2.0 `
`Jzz=.0 `
`basis = spin_basis_1d(L,S='1/2',pauli=False)`
`#basis = spin_basis_1d(L,pauli=False,Nup=L//2)` 
`#basis = spin_basis_1d(L,pauli=False,Nup=L//2,pblock=1)` 
`J_zz = [[Jzz,i,(i+1)%L] for i in range(L-1)] `
`J_xx = [[Jxx,i,(i+1)%L] for i in range(L-1)] `
`J_yy = [[Jyy,i,(i+1)%L] for i in range(L-1)] `
`h_z=[[hz,i] for i in range(L)]`
`static = [["xx",J_xx],["yy",J_yy],["zz",J_zz],["z",h_z]]`
`dynamic=[]`
`H_XYZ = hamiltonian(static,dynamic,basis=basis,dtype=np.float32,check_symm=False)`
`Emin,Emax=H_XYZ.eigsh(k=2,which="BE",maxiter=1E4,return_eigenvectors=False)`
`print(hz, Emin)`

**This script works well for `S=1/2` in `basis`:**
 `Hermiticity check passed!`

`2.0 -10.000003`

**but replacing it by `S=1` gives following errors:**

`Hermiticity check passed!`
`Traceback (most recent call last):`
 ` File "test-uniform.py", line 20, in <module>`
 `   H_XYZ = hamiltonian(static,dynamic,basis=basis,dtype=np.float32,check_symm=False)`
 ` File "/home/user/anaconda3/lib/python3.8/site-packages/quspin/operators/hamiltonian_core.py", line 322, in __init__`
  `  self._static=make_static(self._basis,static_opstr_list,dtype)`
  `File "/home/user/anaconda3/lib/python3.8/site-packages/quspin/operators/_make_hamiltonian.py", line 109, in make_static`
  `  return basis._make_matrix(static_list,dtype)`
  `File "/home/user/anaconda3/lib/python3.8/site-packages/quspin/basis/base.py", line 226, in _make_matrix`
  `  ME,row,col = self.Op(opstr,indx,J,dtype)`
  `File "/home/user/anaconda3/lib/python3.8/site-packages/quspin/basis/base.py", line 218, in Op`
  `  return self._Op(opstr,indx,J,dtype)`
  `File "/home/user/anaconda3/lib/python3.8/site-packages/quspin/basis/basis_1d/spin.py", line 238, in _Op`
  `  ME,row,col = basis_1d._Op(self,opstr,indx,J,dtype)`
  `File "/home/user/anaconda3/lib/python3.8/site-packages/quspin/basis/basis_1d/base_1d.py", line 580, in _Op`
  `  raise ValueError("unrecognized characters {} in operator string.".format(extra_ops))`
`ValueError: unrecognized characters {'x'} in operator string.`

Kindly help me in resolving the issue ?

**Comment by JiahaoYao** [comment, marked as answer]
Hi @sachin00-github , thanks for your question! One note [here](https://weinbe58.github.io/QuSpin/generated/quspin.basis.spin_basis_1d.html#quspin.basis.spin_basis_1d) (2nd in the `Note`)

> Higher spins can only be defined using the spin operators, and do NOT support the operator strings “x” and “y”.

Sorry for the inconvenience, and I am afraid that you have to work out a little bit and express the hamiltonian in terms of `+`, `-` and `z` operator.

**Reply by JiahaoYao** [reply]
I mean, you can use the following relation: 

```markdown
S^x = 1/2 (S^+ + S^-) and S^y = 1/(2i) (S^+ - S^-)
```

**Reply by magnon00** [reply]
Yes, Thanks
